File: src/CarTracker.java
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;
import javafx.stage.Stage;
import java.util.Locale;
import java.util.Scanner;

public class CarTracker extends Subscriber {
    private Stage stage;
    private Label telemetryLabel;
    private Circle carRepresentation;
    private Pane drawingPane;

    private static final double DRAWING_PANE_WIDTH = 500.0; // Ancho del área de rastreo
    private static final double DRAWING_PANE_HEIGHT = 500.0; // Alto del área de rastreo
    private static final double CAR_RADIUS = 8.0; // Radio del círculo que representa el auto

    public CarTracker(String name, String topicName) {
        super(name, topicName);
        stage = new Stage();
        stage.setTitle("Car Tracker: " + name + ", Topic: " + topicName);

        telemetryLabel = new Label("Esperando datos de posición...");
        telemetryLabel.setStyle("-fx-font-size: 14px; -fx-padding: 10px;");

        // Inicializa el círculo que representa el auto
        carRepresentation = new Circle(CAR_RADIUS, Color.RED);
        carRepresentation.setCenterX(CAR_RADIUS);
        carRepresentation.setCenterY(CAR_RADIUS);

        // Panel para dibujar el círculo
        drawingPane = new Pane();
        drawingPane.setPrefSize(DRAWING_PANE_WIDTH, DRAWING_PANE_HEIGHT);
        drawingPane.setStyle("-fx-background-color: white; -fx-border-color: lightgrey;"); // Fondo blanco y borde
        drawingPane.getChildren().add(carRepresentation); // Agrega el círculo al panel de dibujo

        BorderPane rootPane = new BorderPane();
        rootPane.setCenter(drawingPane); // El panel de dibujo en el centro
        rootPane.setBottom(telemetryLabel);
        BorderPane.setAlignment(telemetryLabel, Pos.CENTER);

        // Ajustar el tamaño de la escena para que quepa el panel de dibujo y la etiqueta
        Scene scene = new Scene(rootPane, DRAWING_PANE_WIDTH, DRAWING_PANE_HEIGHT + 50); // 50px extra para la etiqueta
        stage.setScene(scene);
        stage.setResizable(false);
    }

    @Override
    public void update(String message) {
        try (Scanner scanner = new Scanner(message)) {
            scanner.useLocale(Locale.US);
            if (scanner.hasNextDouble()) {
                double time = scanner.nextDouble();
                if (scanner.hasNextDouble()) {
                    double posX = scanner.nextDouble();
                    if (scanner.hasNextDouble()) {
                        double posY = scanner.nextDouble();
                        telemetryLabel.setText(String.format(Locale.US, "t: %.1f, x: %.1f, y: %.1f", time, posX, posY));

                        // Actualiza la posición del círculo en el panel de dibujo [cite: 114]
                        // Se asume que las coordenadas (posX, posY) están dentro de [0, DRAWING_PANE_WIDTH/HEIGHT]
                        // Si posX o posY son mayores que las dimensiones del panel, el círculo podría no ser visible.
                        // No se realiza escalado aquí, se asume mapeo directo.
                        carRepresentation.setCenterX(posX);
                        carRepresentation.setCenterY(posY);

                        // System.out.println(getName() + " actualizado a: t=" + time + ", x=" + posX + ", y=" + posY);
                        return;
                    }
                }
            }
            // Si el formato del mensaje no es el esperado
            System.err.println("CarTracker " + getName() + ": Formato de mensaje incorrecto: " + message);
            telemetryLabel.setText("Error en datos: " + message);
        } catch (Exception e) {
            System.err.println("CarTracker " + getName() + ": Error al procesar mensaje: " + message + " - " + e.getMessage());
            telemetryLabel.setText("Error al procesar: " + message);
        }
    }

    public void show() {
        stage.show();
    }

    public void close() {
        stage.close();
    }
}

File: src/CarTracker.cpp
#include "CarTracker.h"
#include <QVBoxLayout>
#include <QString>
#include <QStringList>
#include <QLabel>
#include <QGraphicsScene>
#include <QGraphicsView>
#include <QGraphicsEllipseItem>
#include <QBrush>

CarTracker::CarTracker(const std::string& name, const std::string& topic, QWidget *parent)
        : Component(name, topic),
          QWidget(parent, Qt::Window),
          Subscriber(name, topic)
{
    // Configuración de la ventana
    setWindowTitle(QString::fromStdString("Car Tracker: " + name + ", Topic: " + topic));

    scene = new QGraphicsScene(this);
    scene->setSceneRect(0, 0, 500, 500);

    view = new QGraphicsView(scene, this);
    view->setBackgroundBrush(QBrush(Qt::white)); // Fondo blanco
    view->setRenderHint(QPainter::Antialiasing); // Para que el círculo se vea suave

    // Creación del círculo que representa el auto
    const double R = 8.0; // Radio del círculo
    carCircle = new QGraphicsEllipseItem(0, 0, 2 * R, 2 * R);
    carCircle->setPos(0, 0); // Posición inicial
    carCircle->setBrush(Qt::red); // Color rojo
    carCircle->setPen(Qt::NoPen); // Sin borde
    scene->addItem(carCircle);

    telemetryLabel = new QLabel("Esperando datos de posición...", this);
    telemetryLabel->setAlignment(Qt::AlignCenter);
    telemetryLabel->setStyleSheet("font-size: 14px; padding: 5px; background-color: #f0f0f0;");

    // Configura el layout
    auto* layout = new QVBoxLayout(this);
    layout->addWidget(view); // La vista gráfica ocupa la mayor parte del espacio
    layout->addWidget(telemetryLabel); // La etiqueta va abajo
    layout->setContentsMargins(0,0,0,0);
    layout->setSpacing(0);
    setLayout(layout);

    resize(502, 550); // Ajustar tamaño de la ventana
}

void CarTracker::update(const std::string& message) {
    QString qMessage = QString::fromStdString(message);
    QStringList parts = qMessage.split(' ');

    if (parts.size() >= 3) {
        bool ok1, ok2, ok3;
        double time = parts[0].toDouble(&ok1);
        double x = parts[1].toDouble(&ok2);
        double y = parts[2].toDouble(&ok3);

        if (ok1 && ok2 && ok3) {
            QString labelText = QString::asprintf("t: %.1f, x: %.1f, y: %.1f", time, x, y);
            telemetryLabel->setText(labelText);

            // El origen (0,0) del círculo es su esquina superior izquierda, se le resta el radio para que el 'centro' del círculo quede en (x,y).
            const double R = carCircle->boundingRect().width() / 2.0;
            carCircle->setPos(x - R, y - R);
        }
    }
}